core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -183,26 +175,30 @@ synchronized void waitForOpsToComplete(final long seqNo) throws InterruptedExcep
     @SuppressForbidden(reason = "Object#notifyAll")
     private void updateCheckpoint() {
         assert Thread.holdsLock(this);
-        assert checkpoint < firstProcessedSeqNo + bitArraysSize - 1 :
-            "checkpoint should be below the end of the first bit set (o.w. current bit set is completed and shouldn't be there)";
-        assert getBitSetForSeqNo(checkpoint + 1) == processedSeqNo.getFirst() :
-            "checkpoint + 1 doesn't point to the first bit set (o.w. current bit set is completed and shouldn't be there)";
-        assert getBitSetForSeqNo(checkpoint + 1).get(seqNoToBitSetOffset(checkpoint + 1)) :
+        assert getBitArrayForSeqNo(checkpoint + 1).get(seqNoToBitArrayOffset(checkpoint + 1)) :
             "updateCheckpoint is called but the bit following the checkpoint is not set";
         try {
             // keep it simple for now, get the checkpoint one by one; in the future we can optimize and read words
-            FixedBitSet current = processedSeqNo.getFirst();
+            long bitArrayKey = getBitArrayKey(checkpoint);
+            FixedBitSet current = processedSeqNo.get(bitArrayKey);
+            if (current == null) {
```

Comment:
when can this happen? it seems we only clean bit sets here?

Reply:
When we clean the set (e.g., the checkpoint is equal to `size - 1`) but the checkpoint can not advance to `size`. In this case, when `size` is marked as completed, checkpoint will still be equal to `size - 1` but we have already cleaned the set.